File: app/src/androidTest/java/at/priv/graf/zazentimer/SettingsTest.java
package at.priv.graf.zazentimer;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.SystemClock;
import android.view.MotionEvent;
import android.widget.SeekBar;

import androidx.preference.PreferenceManager;
import androidx.test.espresso.Espresso;
import androidx.test.espresso.UiController;
import androidx.test.espresso.ViewAction;
import androidx.test.espresso.intent.Intents;
import androidx.test.ext.junit.rules.ActivityScenarioRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.LargeTest;
import androidx.test.platform.app.InstrumentationRegistry;

import org.hamcrest.Matcher;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import dagger.hilt.android.testing.HiltAndroidRule;
import dagger.hilt.android.testing.HiltAndroidTest;
import at.priv.graf.zazentimer.screens.MainPage;
import at.priv.graf.zazentimer.screens.SettingsPage;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.scrollTo;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.intent.Intents.intended;
import static androidx.test.espresso.intent.matcher.IntentMatchers.hasAction;
import static androidx.test.espresso.matcher.ViewMatchers.isAssignableFrom;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withClassName;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

@HiltAndroidTest
@RunWith(AndroidJUnit4.class)
@LargeTest
public class SettingsTest {

    @Rule(order = 0)
    public HiltAndroidRule hiltRule = new HiltAndroidRule(this);

    @Rule(order = 1)
    public ActivityScenarioRule<ZazenTimerActivity> activityRule =
            new ActivityScenarioRule<>(ZazenTimerActivity.class);

    @Before
    public void init() {
        hiltRule.inject();
    }

    @Test
    public void testOpenSettings() {
        new MainPage()
                .verifyMainScreenIsDisplayed()
                .clickToolbarOverflowItem(R.string.menu_settings);
        new SettingsPage();
        new SettingsPage().goBack()
                .verifyMainScreenIsDisplayed();
    }

    @Test
    public void testThemeToggle() {
        new MainPage()
                .verifyMainScreenIsDisplayed()
                .clickToolbarOverflowItem(R.string.menu_settings);
        new SettingsPage();

        onView(withText(R.string.theme)).perform(click());
        onView(withText(R.string.theme_dark)).perform(click());

        new SettingsPage();
    }

    @Test
    public void testMuteSettings() {
        new MainPage()
                .verifyMainScreenIsDisplayed()
                .clickToolbarOverflowItem(R.string.menu_settings);
        new SettingsPage();

        onView(withText(R.string.pref_mute_mode_vibrate_sound))
                .perform(scrollTo(), click());

        SharedPreferences prefs = getPrefs();
        assertTrue(prefs.getBoolean("mute_mode_vibrate_sound", false));
        assertFalse(prefs.getBoolean("mute_mode_none", true));

        onView(withText(R.string.pref_mute_mode_none))
                .perform(scrollTo(), click());

        assertTrue(prefs.getBoolean("mute_mode_none", false));
        assertFalse(prefs.getBoolean("mute_mode_vibrate_sound", true));
    }

    @Test
    public void testBrightnessAdjustment() {
        new MainPage()
                .verifyMainScreenIsDisplayed()
                .clickToolbarOverflowItem(R.string.menu_settings);
        new SettingsPage();

        onView(withText(R.string.checkbox_keep_screen_on))
                .perform(scrollTo(), click());

        onView(withText(R.string.pref_title_brightness))
                .perform(click());

        onView(isAssignableFrom(SeekBar.class))
                .perform(setSeekBarProgress(50));

        Espresso.pressBack();
    }

    @RequiresDisplay
    @Test
    public void testBackup() {
        Intents.init();
        try {
            new MainPage()
                    .verifyMainScreenIsDisplayed()
                    .clickToolbarOverflowItem(R.string.menu_settings);
            new SettingsPage()
                    .clickBackup();

            intended(hasAction(Intent.ACTION_CREATE_DOCUMENT));
        } finally {
            Intents.release();
        }
    }

    @RequiresDisplay
    @Test
    public void testRestore() {
        Intents.init();
        try {
            new MainPage()
                    .verifyMainScreenIsDisplayed()
                    .clickToolbarOverflowItem(R.string.menu_settings);
            new SettingsPage()
                    .clickRestoreAndConfirm();

            intended(hasAction(Intent.ACTION_OPEN_DOCUMENT));
        } finally {
            Intents.release();
        }
    }

    private SharedPreferences getPrefs() {
        return PreferenceManager.getDefaultSharedPreferences(
                InstrumentationRegistry.getInstrumentation().getTargetContext());
    }

    private static ViewAction setSeekBarProgress(final int progress) {
        return new ViewAction() {
            @Override
            public Matcher<android.view.View> getConstraints() {
                return isAssignableFrom(SeekBar.class);
            }

            @Override
            public String getDescription() {
                return "Set SeekBar progress to " + progress;
            }

            @Override
            public void perform(UiController uiController, android.view.View view) {
                SeekBar seekBar = (SeekBar) view;
                int range = seekBar.getWidth()
                        - seekBar.getPaddingLeft() - seekBar.getPaddingRight();
                float target = (float) progress / seekBar.getMax();
                float xPos = seekBar.getPaddingLeft() + target * range;
                float yPos = seekBar.getHeight() / 2f;

                long downTime = SystemClock.uptimeMillis();
                seekBar.dispatchTouchEvent(MotionEvent.obtain(
                        downTime, downTime, MotionEvent.ACTION_DOWN, xPos, yPos, 0));
                seekBar.dispatchTouchEvent(MotionEvent.obtain(
                        downTime, downTime + 50, MotionEvent.ACTION_MOVE, xPos, yPos, 0));
                seekBar.dispatchTouchEvent(MotionEvent.obtain(
                        downTime, downTime + 100, MotionEvent.ACTION_UP, xPos, yPos, 0));
            }
        };
    }
}

File: app/src/androidTest/kotlin/at/priv/graf/zazentimer/SettingsTest.kt
package at.priv.graf.zazentimer

import android.content.Intent
import android.os.SystemClock
import android.view.MotionEvent
import android.widget.SeekBar
import androidx.test.espresso.Espresso
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.UiController
import androidx.test.espresso.ViewAction
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.action.ViewActions.scrollTo
import androidx.test.espresso.intent.Intents
import androidx.test.espresso.intent.Intents.intended
import androidx.test.espresso.intent.matcher.IntentMatchers.hasAction
import androidx.test.espresso.matcher.ViewMatchers.isAssignableFrom
import androidx.test.espresso.matcher.ViewMatchers.withText
import androidx.test.ext.junit.rules.ActivityScenarioRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.LargeTest
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.uiautomator.UiDevice
import androidx.test.uiautomator.UiSelector
import at.priv.graf.zazentimer.screens.MainPage
import at.priv.graf.zazentimer.screens.SettingsPage
import dagger.hilt.android.testing.HiltAndroidRule
import dagger.hilt.android.testing.HiltAndroidTest
import org.hamcrest.Matcher
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@HiltAndroidTest
@RunWith(AndroidJUnit4::class)
@LargeTest
class SettingsTest {
    @get:Rule(order = 0)
    val hiltRule = HiltAndroidRule(this)

    @get:Rule(order = 1)
    val activityRule = ActivityScenarioRule(ZazenTimerActivity::class.java)

    private lateinit var device: UiDevice

    @Before
    fun init() {
        hiltRule.inject()
        device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
    }

    @Test
    fun testOpenSettings() {
        MainPage()
            .verifyMainScreenIsDisplayed()
            .clickToolbarOverflowItem(R.string.menu_settings)
        SettingsPage()
        SettingsPage()
            .goBack()
            .verifyMainScreenIsDisplayed()
    }

    @Test
    fun testThemeToggle() {
        MainPage()
            .verifyMainScreenIsDisplayed()
            .clickToolbarOverflowItem(R.string.menu_settings)
        SettingsPage()

        clickPreferenceByText(R.string.theme)
        clickPreferenceByText(R.string.theme_dark)

        Espresso.onIdle()

        SettingsPage()
    }

    @Test
    fun testBrightnessAdjustment() {
        MainPage()
            .verifyMainScreenIsDisplayed()
            .clickToolbarOverflowItem(R.string.menu_settings)
        SettingsPage()

        clickPreferenceByText(R.string.checkbox_keep_screen_on)

        clickPreferenceByText(R.string.pref_title_brightness)

        Espresso.onIdle()

        onView(isAssignableFrom(SeekBar::class.java))
            .perform(setSeekBarProgress(50))

        Espresso.onIdle()
        Espresso.pressBack()
    }

    @Test
    fun testBackup() {
        Intents.init()
        try {
            MainPage()
                .verifyMainScreenIsDisplayed()
                .clickToolbarOverflowItem(R.string.menu_settings)
            SettingsPage()
                .clickBackup()

            intended(hasAction(Intent.ACTION_CREATE_DOCUMENT))
        } finally {
            Intents.release()
        }
    }

    @Test
    fun testRestore() {
        Intents.init()
        try {
            MainPage()
                .verifyMainScreenIsDisplayed()
                .clickToolbarOverflowItem(R.string.menu_settings)
            SettingsPage()
                .clickRestoreAndConfirm()

            intended(hasAction(Intent.ACTION_OPEN_DOCUMENT))
        } finally {
            Intents.release()
        }
    }

    private fun clickPreferenceByText(textResId: Int) {
        val text = InstrumentationRegistry.getInstrumentation().targetContext.getString(textResId)
        try {
            onView(withText(textResId)).perform(scrollTo(), click())
        } catch (e: Exception) {
            device.findObject(UiSelector().text(text)).click()
        }
    }

    companion object {
        private fun setSeekBarProgress(progress: Int): ViewAction =
            object : ViewAction {
                override fun getConstraints(): Matcher<android.view.View> = isAssignableFrom(SeekBar::class.java)

                override fun getDescription(): String = "Set SeekBar progress to $progress"

                override fun perform(
                    uiController: UiController,
                    view: android.view.View,
                ) {
                    val seekBar = view as SeekBar
                    val range = seekBar.width - seekBar.paddingLeft - seekBar.paddingRight
                    val target = progress.toFloat() / seekBar.max
                    val xPos = seekBar.paddingLeft + target * range
                    val yPos = seekBar.height / 2f

                    val downTime = SystemClock.uptimeMillis()
                    seekBar.dispatchTouchEvent(
                        MotionEvent.obtain(downTime, downTime, MotionEvent.ACTION_DOWN, xPos, yPos, 0),
                    )
                    seekBar.dispatchTouchEvent(
                        MotionEvent.obtain(downTime, downTime + 50, MotionEvent.ACTION_MOVE, xPos, yPos, 0),
                    )
                    seekBar.dispatchTouchEvent(
                        MotionEvent.obtain(downTime, downTime + 100, MotionEvent.ACTION_UP, xPos, yPos, 0),
                    )
                }
            }
    }
}
